### DISCUSSION TITLE
When tuning angular PIDs, the robot seems to ignore tracking wheel and just spins indefinitely

### DISCUSSION THREAD

**Discussion by ItsRobbieman** [LemLib/LemLib#230] [Q&A]
Basically, I'm tuning my angular settings but the robot appears to ignore the horizontal tracking wheel and just spins at full speed. Ive tested with an IMU and no difference. New brain worked for a while then the issue returned. Only theory is that I haven't added anything to my program that uses internal motor encoders to determine heading, but honestly I don't know what else to do anymore, if anyone has any suggestions it would be much appreciated. (Code won't upload but will try again soon)

After testing I've found something may be incorrect with IME tracking as the printed values change incorrectly with no tracking wheels connected. Any ideas with that?

**Comment by meisZWFLZ** [comment]
1. Could you send your code?
1. Are you printing the odom pose to the brain screen? 
  a. Is it accurate?
1. Do you have 2 vertical tracking wheels, or 2 horizontal tracking wheels?
  If so, they will be prioritized over imu. You should only use one tracking wheel in each direction if you have an imu.

**Reply by ItsRobbieman** [reply]
I’ll attempt to send my code again once I’m at my computer. In the meantime, I’m using 1 horizontal and one vertical wheel. Not using IMU as of now. And no, I am not printing coordinates to my brain, but will test when I get the opportunity

**Reply by ItsRobbieman** [reply]
![image](https://github.com/user-attachments/assets/b1a7d5f5-68b1-4072-af16-8cf71a846a7e)

This is what appears on the brain when printing position. I feel there should be an X value maybe? Also, the 0,0,0 doesn't change when moving and the Y changes when moving both directions

**Reply by ItsRobbieman** [reply]
Upon further testing I've found that theta and Y both change when moving forward and backward with no tracking wheels connected, leading me to believe something with the IMEs is programmed wrong? I'll attach code if that helps

**Reply by meisZWFLZ** [reply]
Were you able to attach your code?

**Reply by meisZWFLZ** [reply]
Also, did you set up driver control with LemLib, and does it work properly?

**Reply by ItsRobbieman** [reply]
Zip file is too large but if there's another way u know of, I'll give it a shot. I've made a second code for simplicity during testing that I'll attach. Just as a run down (sorry I've posted about 6 different things in here by now) I THINK it's ignoring the tracking wheels and only using IMEs, which are incorrect

**Reply by ItsRobbieman** [reply]
And yes, driver control works fine, and always has

**Reply by meisZWFLZ** [reply]
Is your code spread apart many different files or just main.cpp? If its the latter, could you just send main.cpp?
If its in multiple files, you could try to upload it to google drive, or make a github repository and put the folder in there.

**Reply by ItsRobbieman** [reply]
[Link to the file on google drive]
(https://drive.google.com/drive/folders/11OWJqxM8XZjaWXyds30jblbHXhJFobEV?usp=sharing)

**Reply by ItsRobbieman** [reply]
may be a bit messy due to me testing different things, but if u think u can fix it, please do. At this point I believe the IMEs are wrong, so either they need to be fixed or the tracking wheels need to take priority. Also, I believe there's a config issue because when it moves forward, X and theta decrease, and Y increases

**Reply by meisZWFLZ** [reply]
Can you try using lemlib for driver according to the docs:
https://lemlib.readthedocs.io/en/stable/tutorials/3_driver_control.html

This should help to diagnose whats going on. Ie. whether the motors are reversed correctly
You should also be fine to just send the main.cpp file.

**Reply by ItsRobbieman** [reply]
I have driver control working fine, all motors turn correctly. It's in a separate code but it works with no issues, so I'd assume all motors are reversed correctly

**Reply by meisZWFLZ** [reply]
wdym by different code? whats the code?

**Reply by ItsRobbieman** [reply]
Like a completely different project

**Reply by meisZWFLZ** [reply]
can you send that code? Also, in the main.cpp you uploaded to google drive, you specify all the sensors as `nullptr`, so thats why its using imes.

**Comment by ItsRobbieman** [comment]
https://drive.google.com/drive/folders/1HD8hml7cC951wxK_028n1CelZOYJDD-h?usp=sharing.       have tested with sensors added and with nullptr. The values don't seem to change even with tracking wheels configured
